resource/composed/composed.go

```diff
@@ -29,17 +30,68 @@ import (
 	"github.com/crossplane/function-sdk-go/errors"
 )
 
+// Scheme used to determine the type of any runtime.Object passed to From.
+var Scheme *runtime.Scheme
+
+func init() {
+	Scheme = runtime.NewScheme()
+}
+
 // New returns a new unstructured composed resource.
 func New() *Unstructured {
 	return &Unstructured{unstructured.Unstructured{Object: make(map[string]any)}}
 }
 
 // From creates a new unstructured composed resource from the supplied object.
 func From(o runtime.Object) (*Unstructured, error) {
-	obj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(o)
+	// If the supplied object is already unstructured content, avoid a JSON
+	// round trip and use it.
+	if u, ok := o.(interface{ UnstructuredContent() map[string]any }); ok {
+		return &Unstructured{unstructured.Unstructured{Object: u.UnstructuredContent()}}, nil
+	}
+
+	// Set the object's GVK from our scheme.
+	gvks, _, err := Scheme.ObjectKinds(o)
+	if err != nil {
+		return nil, errors.Wrap(err, "did you add it to composed.Scheme?")
+	}
+	// There should almost never be more than one GVK for a type.
+	for _, gvk := range gvks {
+		o.GetObjectKind().SetGroupVersionKind(gvk)
+	}
+
+	// Round-trip the supplied object through JSON to convert it. We use the
+	// go-json-experiment package for this because it honors the omitempty field
+	// for non-pointer struct fields.
+	//
+	// At the time of writing many Crossplane structs contain fields that have
+	// the omitempty struct tag, but non-pointer struct values. pkg/json does
+	// not omit these fields. It instead includes them as empty JSON objects.
+	// Crossplane will interpret this as part of a server-side apply fully
+	// specified intent and assume the function actually has opinion about the
+	// field when it doesn't. We should make these fields pointers, but it's
+	// easier and safer in the meantime to work around it here.
+	//
+	// https://github.com/go-json-experiment/json#behavior-changes
+	j, err := json.Marshal(o)
 	if err != nil {
 		return nil, err
 	}
+	obj := make(map[string]any)
+	if err := json.Unmarshal(j, &obj); err != nil {
+		return nil, err
+	}
+
+	// Unfortunately go-json-experiment also has different omitempty semantics
+	// for integers, in that it will include a zero-value integer with the
+	// omitempty tag. That will be the case for metadata.generation, so we
+	// delete it. It's not something a function should ever be setting.
+	if m, exists := obj["metadata"]; exists {
+		if mo, ok := m.(map[string]any); ok {
+			delete(mo, "generation")
+		}
+	}
+
```

Comment:
feels like we could add a cleanup step to Crossplane then probably

Reply:
You mean do something like this on the Crossplane side just in case? Yeah potentially. Maybe let's wait and see how much of an issue it will be in real life. I thought about addressing it there to start with, but it should only be an issue for Go functions and hopefully they'll all use this SDK...